content/articles/gulp4-primer.md

```diff
@@ -0,0 +1,153 @@
+);
+```
+
+In that case we first want to run build, and then browser-sync, watch and karma-watch in parallel. Easy (the commit below will contain a buggy version since I did put browser-sync in the series by mistake, *easy* right).  
+
+Now trying to run the build task, I get another error:
+
+```js
+[15:09:07] TypeError: undefined is not a function
+    at /home/vincent/Code/ng-boilerplate/node_modules/run-sequence/index.js:18:22
+    at Array.forEach (native)
+    at verifyTaskSets (/home/vincent/Code/ng-boilerplate/node_modules/run-sequence/index.js:12:11)
+```
+
+Since gulp 4 has a method to run things sequentially, let's use it.
+
+```js
+// before
+gulp.task('build', function(){
+  return runSequence(
+    'clean',
+    ['sass', 'copy-assets', 'ts-compile', 'templates', 'copy-vendor'],
+    'index'
+  );
+)};
+
+// after
+gulp.task(
+  'build',
+  gulp.series(
+    'clean',
+    gulp.parallel('sass', 'copy-assets', 'ts-compile', 'templates', 'copy-vendor'),
+    'index'
+  )
+);
+```
+
+> Note: I had to modify some tasks like the browser-sync since its API changed.
+I also had to make a [patch](https://github.com/dlmanning/gulp-sass/pull/207) to gulp-sass to update the node-sass version.  
+
+You can see the commit [here](https://github.com/Keats/ng-boilerplate/commit/91a05401c6ea532467bc00b5a6e54fd95b6b0eaf).  
+
+Looking at the [build](https://travis-ci.org/Keats/ng-boilerplate/builds/54274589), it still fails because of a missing selenium jar which means a protractor error.
+My elegant solution was the [following](https://github.com/Keats/ng-boilerplate/commit/d999db2442ac72fc440e8ff9cccd8348c533c72d), as in removing protractor from travis because I'm not using it right now anyway.
+
+### Small cleanup
+The following is not strictly related to gulp4 but contains some of the things I like to do to have a clean gulpfile.
+The config part at the top of the gulpfile (where I am defining all the paths to inject) is quite ugly but I am not going to touch it here, only change the gulp related things.
+
+First thing, [gulp-load-plugins](https://www.npmjs.com/package/gulp-load-plugins) makes things quite simple but the prefix I had chosen, 'plugins', is quite verbose and takes up space. Recently I've started using `$` which works quite well.
+Something that took me way too long to realise is that it auto-camelCases packages names with dash so my following line was useless:
+
+```js
+plugins.ngAnnotate = require('gulp-ng-annotate');
+```
+
+Another thing to examine is whether some of the gulp plugins you're using actually bring any values compared to using the actual npm module directly.
+In my boilerplate, gulp-karma is not doing anything special and was only necessary while the karma team was fixing some of the issues the plugin was solving.
+
+While going through the file, I also realised I didn't test the watch process at all and that it didn't work (surprising eh).  
```

Comment:
I don't know if this matters in a blog post, but there are two extra whitespaces at the end of the line.